fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java

```diff
@@ -0,0 +1,166 @@
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Range;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Used to prune partition of olap scan, should execute after SwapProjectAndFilter, MergeConsecutiveFilters,
+ * MergeConsecutiveProjects and all predicate push down related rules.
+ */
+public class PruneOlapScanPartition extends OneRewriteRuleFactory {
+
+    @Override
+    public Rule build() {
+        return logicalFilter(logicalOlapScan()).thenApply(ctx -> {
+            LogicalFilter<LogicalOlapScan> filter = ctx.root;
+            LogicalOlapScan scan = filter.child();
+            Expression predicate = filter.getPredicates();
+            List<Expression> expressionList = ExpressionUtils.extractConjunction(predicate);
+            OlapTable table = scan.getTable();
+            Set<String> partitionColumnNameSet = Utils.execWithReturnVal(table::getPartitionColumnNames);
+            PartitionInfo partitionInfo = table.getPartitionInfo();
+            // TODO: 1. support grammar: SELECT * FROM tbl PARTITION(p1,p2)
+            //       2. support list partition
+            if (partitionColumnNameSet.isEmpty() || !partitionInfo.getType().equals(PartitionType.RANGE)) {
+                return ctx.root;
+            }
+            // TODO: Process all partition column for now, better to process required column only.
+            Map<String, ColumnRange> columnNameToRange = Maps.newHashMap();
+            for (String colName : partitionColumnNameSet) {
+                ColumnRange columnRange = createColumnRange(colName, expressionList);
+                columnNameToRange.put(colName, columnRange);
+            }
+
+            Map<Long, PartitionItem> keyItemMap = partitionInfo.getIdToItem(false);
+            PartitionPruner partitionPruner = new RangePartitionPrunerV2(keyItemMap,
+                    partitionInfo.getPartitionColumns(), columnNameToRange);
+            Collection<Long> selectedPartitionId =  Utils.execWithReturnVal(partitionPruner::prune);
+            scan.getSelectedPartitionIds().retainAll(selectedPartitionId);
+            return ctx.root;
+        }).toRule(RuleType.PARTITION_PRUNE);
+    }
+
+    private ColumnRange createColumnRange(String colName, List<Expression> expressionList) {
+        ColumnRange result = ColumnRange.create();
+        for (Expression expression : expressionList) {
+            Expression leftMostChild = expression.leftMostNode();
+            if (!(leftMostChild instanceof SlotReference)) {
+                continue;
+            }
+            SlotReference slotRef = (SlotReference) leftMostChild;
+            if (!slotRef.boundToColumn(colName)) {
+                continue;
+            }
+            if (expression instanceof Or) {
+                ColumnRanges ranges = exprToRanges(expression, colName);
```

Comment:
this is strange that `expression` is a `Or`, but `exprToRanges` need a `ComparisonPredicate `.
furthermore, if and else branches are same.
So, do u forget something?